docs/man/man3/Navigation.3
.TH "Navigation" 3 "Fri Nov 22 2019" "Version 7" "GMock" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Navigation \- Class \fBNavigation\fP Contains the methods of Pid Algorithm\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Navigation\&.hpp>\fP
.PP
Inherits \fBSteerAlgorithm\fP\&.
.PP
Inherited by \fBMockNavigation\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBNavigation\fP ()"
.br
.RI "\fIConstructor of class \fBNavigation\fP\&. \fP"
.ti -1c
.RI "virtual \fB~Navigation\fP ()"
.br
.RI "\fIDestructor of class \fBNavigation\fP\&. \fP"
.ti -1c
.RI "virtual std::vector< double > \fBcalculatePID\fP (double setPoint, double currentVelocity)"
.br
.RI "\fIFunction to calculate new velocity using PID in m/s\&. \fP"
.ti -1c
.RI "bool \fBgnuVelocityGraph\fP (std::vector< std::pair< double, double >> vectorPointsVel, double newVelocity, bool flag)"
.br
.RI "\fIFunction to use GNUplot for velocity convergence graph\&. \fP"
.ti -1c
.RI "bool \fBgnuSteerAngleGraph\fP (std::vector< std::pair< double, double >> vectorPointsSteer, double steerAngle, bool flag)"
.br
.RI "\fIFunction to use GNUplot for Steer angle convergence graph\&. \fP"
.ti -1c
.RI "virtual double \fBcalculate\fP (double \fBtargetHeading\fP, double currentVelocity, double setPoint, int flag)"
.br
.RI "\fIFunction to calculate new velocity in m/s with a PID Algorithm using kp, ki & kd\&. \fP"
.ti -1c
.RI "virtual double \fBgetKp_\fP ()"
.br
.RI "\fIFunction to get kp_\&. \fP"
.ti -1c
.RI "virtual double \fBgetKi_\fP ()"
.br
.RI "\fIFunction to get ki_\&. \fP"
.ti -1c
.RI "virtual double \fBgetKd_\fP ()"
.br
.RI "\fIFunction to get kd_\&. \fP"
.ti -1c
.RI "virtual bool \fBsetKp_\fP (double kp)"
.br
.RI "\fIFunction to set kp_\&. \fP"
.ti -1c
.RI "virtual bool \fBsetKi_\fP (double ki)"
.br
.RI "\fIFunction to set ki_\&. \fP"
.ti -1c
.RI "virtual bool \fBsetKd_\fP (double kd)"
.br
.RI "\fIFunction to set kd_\&. \fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "bool \fBmotorDirection\fP = true"
.br
.in -1c
.SH "Detailed Description"
.PP 
Class \fBNavigation\fP Contains the methods of Pid Algorithm\&. 
.PP
Definition at line 43 of file Navigation\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Navigation::Navigation ()"

.PP
Constructor of class \fBNavigation\fP\&. 
.PP
Definition at line 44 of file Navigation\&.cpp\&.
.SS "Navigation::~Navigation ()\fC [virtual]\fP"

.PP
Destructor of class \fBNavigation\fP\&. 
.PP
Definition at line 54 of file Navigation\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "double Navigation::calculate (double targetHeading, double currentVelocity, double setPoint, int flag)\fC [virtual]\fP"

.PP
Function to calculate new velocity in m/s with a PID Algorithm using kp, ki & kd\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdouble\fP targetHeading, Target heading of the robot 
.br
\fIdouble\fP currentVelocity, current velocity of robot 
.br
\fIdouble\fP setPoint, Target Velocity 
.br
\fIint\fP flag, to enable while 
.RE
.PP
\fBReturns:\fP
.RS 4
double newVelocity 
.RE
.PP

.PP
Definition at line 128 of file Navigation\&.cpp\&.
.SS "std::vector< double > Navigation::calculatePID (double setPoint, double currentVelocity)\fC [virtual]\fP"

.PP
Function to calculate new velocity using PID in m/s\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdouble\fP setPoint, Setpoint of the robot 
.br
\fIdouble\fP currentVelocity, current Velocity of the robot 
.RE
.PP
\fBReturns:\fP
.RS 4
double std::vector<double>, values of error, Proportional output, Integral output, derivative output 
.RE
.PP

.PP
Definition at line 56 of file Navigation\&.cpp\&.
.SS "double Navigation::getKd_ ()\fC [virtual]\fP"

.PP
Function to get kd_\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
double 
.RE
.PP

.PP
Definition at line 241 of file Navigation\&.cpp\&.
.SS "double Navigation::getKi_ ()\fC [virtual]\fP"

.PP
Function to get ki_\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
double 
.RE
.PP

.PP
Definition at line 237 of file Navigation\&.cpp\&.
.SS "double Navigation::getKp_ ()\fC [virtual]\fP"

.PP
Function to get kp_\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
double 
.RE
.PP

.PP
Definition at line 233 of file Navigation\&.cpp\&.
.SS "bool Navigation::gnuSteerAngleGraph (std::vector< std::pair< double, double >> vectorPointsSteer, double steerAngle, bool flag)"

.PP
Function to use GNUplot for Steer angle convergence graph\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstd::vector<std::pair<double,double>>\fP points vector 
.br
\fIdouble\fP steerAngle, steer angle from code 
.br
\fIbool\fP flag, to test the gnuplot working 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
Definition at line 105 of file Navigation\&.cpp\&.
.SS "bool Navigation::gnuVelocityGraph (std::vector< std::pair< double, double >> vectorPointsVel, double newVelocity, bool flag)"

.PP
Function to use GNUplot for velocity convergence graph\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstd::vector<std::pair<double,double>>\fP points vector 
.br
\fIdouble\fP newVelocity, New velocity from code 
.br
\fIbool\fP flag, to test the gnuplot working 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
Definition at line 83 of file Navigation\&.cpp\&.
.SS "bool Navigation::setKd_ (double kd)\fC [virtual]\fP"

.PP
Function to set kd_\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdouble\fP kd 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean true 
.RE
.PP

.PP
Definition at line 263 of file Navigation\&.cpp\&.
.SS "bool Navigation::setKi_ (double ki)\fC [virtual]\fP"

.PP
Function to set ki_\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdouble\fP ki 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean true 
.RE
.PP

.PP
Definition at line 254 of file Navigation\&.cpp\&.
.SS "bool Navigation::setKp_ (double kp)\fC [virtual]\fP"

.PP
Function to set kp_\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdouble\fP kp 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean true 
.RE
.PP

.PP
Definition at line 245 of file Navigation\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "bool Navigation::motorDirection = true"

.PP
Definition at line 49 of file Navigation\&.hpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for GMock from the source code\&.


app/Navigation.cpp

/************************************************************************
MIT License
Copyright © 2019 Raj Shinde
Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the “Software”), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software,
and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:
The above copyright notice and this permission notice shall be included 
in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
 *************************************************************************/

/**
 *  @copyright MIT License, Copyright © 2019 Raj Shinde
 *  @file    Navigation.cpp
 *  @author  Sprint-1 Raj Shinde- driver and Prasheel Renkuntla- navigator
 *  @author  Sprint-2 Prasheel Renkuntla- driver and Raj Shinde- navigator
 *  @author  Gmock Prasheel Renkuntla
 *  @date    11/22/2019
 *  @version 7.0
 *  @brief   Mid Term Project
 *  @section Implements Ackermann on PID control
 */
#include <time.h>

#include <iostream>
#include <fstream>
#include <vector>

#include "gnuplot-iostream.h"
#include "Navigation.hpp"
#include "SteerAlgorithm.hpp"

Navigation::Navigation() {
//  Initialised Variables
kp_ = 0.5;
ki_ = 0.05;
kd_ = 0.001;
diffTime_ = 0.002;
error_ = 0;
previousError_ = 0;
}

Navigation::~Navigation() {}

std::vector<double> Navigation::calculatePID(double setPoint,
                                      double currentVelocity) {
std::vector<double> pidOut;
double errorValue = 0;     //  to store current error
double propOutput = 0;     //  to store proportional output
double integralVal = 0;    //  temp variable to calculate integral value
double intOutput = 0;      //  to store integral output
double derivativeVal = 0;  //  temp variable to calculate derivative value
double derOutput = 0;      //  to store derivative output

errorValue = setPoint - currentVelocity;
propOutput = kp_ * errorValue;

integralVal += errorValue * diffTime_;
intOutput = ki_ * integralVal;

derivativeVal = ((errorValue - previousError_) / diffTime_);
derOutput = kd_ * derivativeVal;

pidOut.push_back(errorValue);
pidOut.push_back(propOutput);
pidOut.push_back(intOutput);
pidOut.push_back(derOutput);
return pidOut;
}

bool Navigation::gnuVelocityGraph(std::vector<std::pair<double, double>>
                         vectorPointsVel, double newVel, bool flag) {
//  call object and intialise variables for Graph in GNUPLOT
Gnuplot gnup;
//  set the graph on gnuplot for the respective coordinates
gnup << "set xrange [0:0.03]\nset yrange [0:45]\n";
gnup << "set title \"Velocity Convergence\"\n";
gnup << "set pointsize 1\n";
gnup << "set xlabel \"Time\"\n";
gnup << "set ylabel \"Current Velocity\"\n";
gnup << "set key outside\n";

if (flag) {
gnup << "plot" << gnup.file1d(vectorPointsVel)
     << "with lp title 'Current Velocity' lc 3, "
     << newVel << " title 'Set Point' lt 1 lc 4" << std::endl;
} else {
gnup << "test" << std::endl;
}
return flag;
}

bool Navigation::gnuSteerAngleGraph(std::vector<std::pair<double, double>>
                   vectorPointsSteer, double steerAngle, bool flag) {
//  call object and intialise variables for Graph in GNUPLOT
Gnuplot gnu;

//  set the graph on gnuplot for the respective coordinates
gnu << "set xrange [0:2]\nset yrange [0:100]\n";
gnu << "set title \"Steering angle Convergence\"\n";
gnu << "set pointsize 1\n";
gnu << "set xlabel \"Time\"\n";
gnu << "set ylabel \"Heading Angle\"\n";
gnu << "set key outside\n";

if (flag) {
gnu << "plot" << gnu.file1d(vectorPointsSteer)
    << "with points title 'Heading' lc 3, "
    << steerAngle << " title 'Target Heading' lt 1 lc 4" << std::endl;
} else {
gnu << "test" << std::endl;
}
return flag;
}

double Navigation::calculate(double targetHeading,
                             double currentVelocity,
                             double setPoint,
                             int flag) {
double newVelocity = 0;
double tempVel;            //  temp variable to store current velocity
int velocityConverged = 0;
int headingConverged = 0;
std::vector<double> pidOutput;

double outTime = 0;        //  temp variable to determine convergence time
SteerAlgorithm Ackermann = SteerAlgorithm();

double tempHeading = targetHeading;      //  to store the target for graph

//  std::vector<std::pair<double, double>> combinedXY(1, std::make_pair(0, 0));
std::vector<std::pair<double, double>> points;
std::vector<std::pair<double, double>> pointsVelocity;

//  while loop to run until velocity and heading are converged
while(velocityConverged != 1 && headingConverged != 1) {
        //  Run PID control initially
        pidOutput = calculatePID(setPoint, currentVelocity);
        newVelocity = currentVelocity +
                      (pidOutput[1] + pidOutput[2] + pidOutput[2]);
        previousError_ = pidOutput[0];
        currentVelocity = newVelocity;

        clock_t toc = clock();

        //  store the current values in a vector for plotting
        pointsVelocity.push_back(std::make_pair((
                                 static_cast<double> (toc)/CLOCKS_PER_SEC),
                                 currentVelocity));
        if (flag == 2) {
        //  show the values on screen
        std::cout << "Current Velocity: " << currentVelocity
                  << " Setpoint: " << newVelocity << std::endl;
        }
        if (setPoint == newVelocity) {
           velocityConverged = 1;
        }
        if ((targetHeading-heading) != 0 && velocityConverged == 1) {
          if (newVelocity > maxTurnVelocity) {
             tempVel = maxTurnVelocity;
          }
          tempVel = newVelocity;
          if (targetHeading-heading > 0) {
             dir = 1;
          } else {
             dir = -1;
          }
          //  change wheel angles to move robot towards target
          Ackermann.changeWheelAngles(Ackermann.getCorrRadius_(),
                                      shaftLength,
                                      shaftDistance);

          double arcLen = Ackermann.arcLength(targetHeading,
                                      Ackermann.getCorrRadius_());
          auto turnTime = Ackermann.turnTime(arcLen, newVelocity);
          clock_t tic = clock();

          //  loop until heading is converged towards target
          while ((outTime < turnTime) && (targetHeading != heading)) {
                pidOutput = calculatePID(setPoint, currentVelocity);
                newVelocity = currentVelocity +
                             (pidOutput[1] + pidOutput[2] + pidOutput[2]);
                previousError_ = pidOutput[0];
                currentVelocity = newVelocity;

                clock_t toc = clock();
                outTime = (static_cast<double> (toc - tic)) / CLOCKS_PER_SEC;
                double newArcLen = tempVel * outTime;
                heading = (newArcLen * targetHeading / arcLen);

                //  store the current values in a vector for plotting
                points.push_back(std::make_pair(
                                  (static_cast<double> (toc)/CLOCKS_PER_SEC),
                                   heading));
                if (flag == 2) {
                std::cout << "Current Velocity: " << currentVelocity
                          << " Setpoint: " << newVelocity << std::endl;
                std::cout << "Current Heading: " << heading
                          << " Target: " << targetHeading << std::endl;
                }
          }
        targetHeading = 0;
        headingConverged = 1;
        Ackermann.resetWheel();
        }
}

if (flag == 2) {
  //  Plot the graph on gnuplot by calling the below lines
  gnuSteerAngleGraph(points, tempHeading, true);
  gnuVelocityGraph(pointsVelocity, newVelocity, true);
return 0;
} else if (flag == 1) {
  return newVelocity;
} else {
  return heading;
}
return newVelocity;
}

double Navigation::getKp_() {
return kp_;
}

double Navigation::getKi_() {
return ki_;
}

double Navigation::getKd_() {
return kd_;
}

bool Navigation::setKp_(double kp) {
bool flag = true;
kp_ = kp;
if (kp_ != kp) {
    flag = false;
}
return flag;
}

bool Navigation::setKi_(double ki) {
bool flag = true;
ki_ = ki;
if (ki_ != ki) {
    flag = false;
}
return flag;
}

bool Navigation::setKd_(double kd) {
bool flag = true;
kd_ = kd;
if (kd_ != kd) {
    flag = false;
}
return flag;
}
